File: vandal/server/mural.js
"use strict";

// =============================================================================
// mural.js — Authoritative server-side mural state: stroke history, live
//            open strokes, commit/undo logic. Pixel-agnostic (no rasterization).
// Estado autoritativo del mural en el servidor: historial de trazos, trazos
// abiertos en vivo, logica de commit/undo. Agnostico a pixeles (sin rasterizacion).
//
// Key concepts / Conceptos clave:
//   - Ordered list of committed strokes (the full mural for new joiners) / Lista ordenada de trazos confirmados (el mural completo para nuevos usuarios)
//   - Map of open (in-progress) strokes, one per owner / Mapa de trazos abiertos (en progreso), uno por propietario
//   - Auto-capped history (MAX_STROKES) to bound memory / Historial con limite automatico (MAX_STROKES) para acotar la memoria
//   - Undo removes the most recent stroke by owner ID / Deshacer elimina el trazo mas reciente por ID de propietario
// =============================================================================

// ---------------------------------------------------------------------------
// mural.js — the authoritative shared canvas state.
//
// The server is a pixel-agnostic historian: it never rasterizes anything. It
// keeps (a) an ordered list of completed STROKE records so any new joiner can
// be handed the whole mural and undo can target a stroke by id, and (b) a map
// of OPEN strokes still being streamed live (one per owner). Rendering lives
// entirely on the clients.
// ---------------------------------------------------------------------------

const C = require("./constants.js");

function clampInt(v, lo, hi) {
  v = Math.round(v);
  return v < lo ? lo : v > hi ? hi : v;
}

const SHAPE_TOOLS = new Set([1, 2, 3]); // line / rect / circle need 2 points

class Mural {
  constructor() {
    this.strokes = []; // ordered oldest -> newest (completed)
    this.open = new Map(); // ownerId -> open stroke being streamed
    this.nextStrokeId = 1;
  }

  _norm(raw) {
    return {
      tool: clampInt(raw.tool || 0, 0, C.TOOL_COUNT - 1),
      color: clampInt(raw.color || 0, 0, C.PALETTE_COUNT - 1),
      size: clampInt(raw.size || 0, 0, C.SIZE_COUNT - 1),
      flags: (raw.flags | 0) & 0xff,
    };
  }

  _pushCapped(stroke) {
    this.strokes.push(stroke);
    if (this.strokes.length > C.MAX_STROKES) {
      this.strokes.splice(0, this.strokes.length - C.MAX_STROKES);
    }
  }

  // ---- one-shot completed stroke (shapes + eraser) ------------------------
  // Returns the stored stroke, or null if it carried no usable geometry.
  commitStroke(ownerId, raw) {
    const m = this._norm(raw);
    let pts = Array.isArray(raw.points) ? raw.points : [];
    if (pts.length > C.MAX_POINTS) pts = pts.slice(0, C.MAX_POINTS);
    const points = [];
    for (const p of pts) {
      points.push({ x: clampInt(p.x, 0, C.CANVAS_W), y: clampInt(p.y, 0, C.CANVAS_H) });
    }
    if (points.length === 0) return null;
    if (SHAPE_TOOLS.has(m.tool) && points.length < 2) return null; // degenerate

    const stroke = { id: this.nextStrokeId++, ownerId: ownerId >>> 0, ...m, points };
    this._pushCapped(stroke);
    return stroke;
  }

  // ---- streaming (brush + spray) ------------------------------------------
  begin(ownerId, raw, x, y) {
    const m = this._norm(raw);
    const stroke = {
      id: this.nextStrokeId++,
      ownerId: ownerId >>> 0,
      ...m,
      points: [{ x: clampInt(x, 0, C.CANVAS_W), y: clampInt(y, 0, C.CANVAS_H) }],
    };
    this.open.set(ownerId, stroke);
    return stroke;
  }

  // Append points into an owner's open stroke, up to the per-stroke cap.
  // Returns the points actually appended (clamped) and whether the stroke is
  // now full (the caller then auto-splits to keep the line unbroken).
  append(ownerId, pts) {
    const open = this.open.get(ownerId);
    if (!open) return { appended: [], full: false };
    const appended = [];
    for (const p of pts) {
      if (open.points.length >= C.MAX_POINTS) break;
      const q = { x: clampInt(p.x, 0, C.CANVAS_W), y: clampInt(p.y, 0, C.CANVAS_H) };
      open.points.push(q);
      appended.push(q);
    }
    return { appended, full: open.points.length >= C.MAX_POINTS };
  }

  isOpen(ownerId) {
    return this.open.has(ownerId);
  }

  // Finalize an owner's open stroke into history. Returns it, or null.
  end(ownerId) {
    const open = this.open.get(ownerId);
    if (!open) return null;
    this.open.delete(ownerId);
    if (open.points.length === 0) return null;
    this._pushCapped(open);
    return open;
  }

  // Remove the most recent completed stroke authored by ownerId. Returns id/0.
  undoLast(ownerId) {
    for (let i = this.strokes.length - 1; i >= 0; i--) {
      if (this.strokes[i].ownerId === ownerId) {
        const id = this.strokes[i].id;
        this.strokes.splice(i, 1);
        return id;
      }
    }
    return 0;
  }

  removeById(id) {
    for (let i = this.strokes.length - 1; i >= 0; i--) {
      if (this.strokes[i].id === id) {
        this.strokes.splice(i, 1);
        return true;
      }
    }
    return false;
  }
}

module.exports = { Mural };

File: vandal/server-python/mural.py
"""mural.py — the authoritative shared canvas state (Python port).

Mirror of ../server/mural.js. The server is a pixel-agnostic historian: it
never rasterizes anything. It keeps (a) an ordered list of completed STROKE
records so any new joiner can be handed the whole mural and undo can target a
stroke by id, and (b) a map of OPEN strokes still being streamed live (one per
owner). Rendering lives entirely on the clients.
"""

import math

import constants as C

SHAPE_TOOLS = {1, 2, 3}  # line / rect / circle need 2 points


def clamp_int(v, lo, hi):
    v = math.floor(v + 0.5)  # JS Math.round
    return lo if v < lo else (hi if v > hi else v)


class Mural:
    def __init__(self):
        self.strokes = []        # ordered oldest -> newest (completed)
        self.open = {}           # ownerId -> open stroke being streamed
        self.next_stroke_id = 1

    def _norm(self, raw):
        return {
            "tool": clamp_int(raw.get("tool") or 0, 0, C.TOOL_COUNT - 1),
            "color": clamp_int(raw.get("color") or 0, 0, C.PALETTE_COUNT - 1),
            "size": clamp_int(raw.get("size") or 0, 0, C.SIZE_COUNT - 1),
            "flags": (int(raw.get("flags") or 0)) & 0xFF,
        }

    def _push_capped(self, stroke):
        self.strokes.append(stroke)
        if len(self.strokes) > C.MAX_STROKES:
            del self.strokes[0:len(self.strokes) - C.MAX_STROKES]

    # ---- one-shot completed stroke (shapes + eraser) --------------------
    def commit_stroke(self, owner_id, raw):
        m = self._norm(raw)
        pts = raw.get("points") if isinstance(raw.get("points"), list) else []
        if len(pts) > C.MAX_POINTS:
            pts = pts[:C.MAX_POINTS]
        points = []
        for p in pts:
            points.append({"x": clamp_int(p["x"], 0, C.CANVAS_W),
                           "y": clamp_int(p["y"], 0, C.CANVAS_H)})
        if len(points) == 0:
            return None
        if m["tool"] in SHAPE_TOOLS and len(points) < 2:
            return None
        stroke = {"id": self.next_stroke_id, "ownerId": owner_id & 0xFFFFFFFF,
                  **m, "points": points}
        self.next_stroke_id += 1
        self._push_capped(stroke)
        return stroke

    # ---- streaming (brush + spray) --------------------------------------
    def begin(self, owner_id, raw, x, y):
        m = self._norm(raw)
        stroke = {"id": self.next_stroke_id, "ownerId": owner_id & 0xFFFFFFFF,
                  **m, "points": [{"x": clamp_int(x, 0, C.CANVAS_W),
                                   "y": clamp_int(y, 0, C.CANVAS_H)}]}
        self.next_stroke_id += 1
        self.open[owner_id] = stroke
        return stroke

    def append(self, owner_id, pts):
        open_stroke = self.open.get(owner_id)
        if not open_stroke:
            return [], False
        appended = []
        for p in pts:
            if len(open_stroke["points"]) >= C.MAX_POINTS:
                break
            q = {"x": clamp_int(p["x"], 0, C.CANVAS_W),
                 "y": clamp_int(p["y"], 0, C.CANVAS_H)}
            open_stroke["points"].append(q)
            appended.append(q)
        return appended, len(open_stroke["points"]) >= C.MAX_POINTS

    def is_open(self, owner_id):
        return owner_id in self.open

    def end(self, owner_id):
        open_stroke = self.open.get(owner_id)
        if not open_stroke:
            return None
        del self.open[owner_id]
        if len(open_stroke["points"]) == 0:
            return None
        self._push_capped(open_stroke)
        return open_stroke

    def undo_last(self, owner_id):
        for i in range(len(self.strokes) - 1, -1, -1):
            if self.strokes[i]["ownerId"] == owner_id:
                sid = self.strokes[i]["id"]
                del self.strokes[i]
                return sid
        return 0

    def remove_by_id(self, sid):
        for i in range(len(self.strokes) - 1, -1, -1):
            if self.strokes[i]["id"] == sid:
                del self.strokes[i]
                return True
        return False
